### PR TITLE
V4.1 update

### PR DESCRIPTION
**Pull request by ajgpitch**
This version now uses C++ compilation rather than C. This should allow compilation under MSVC14. Therefore all Python versions should be compatible.

For some reason mingw must still be used for Python 2.7 and 3.4. So the compilation must be conditional, this has made the recipe a bit tricky, but let's see how the tests go.

### CODE REVIEW

**Comment by sahmed95** [comment]
Hi, the Appveyor build failure seems to be some issue in parsing the .yml file related to mingwpy. 
```
mingwpy # [win and (py2 or py34)]
```
Maybe a comibined and/or statement isn't valid.
I raised this question in the conda-forge gitter channel and will post here as soon as I get a reply. Running on Windows seems like such a pain. Will shifiting to py3.5+ and using the msvc compiler solve these issues for good ?

**Comment by sahmed95** [comment]
Hi, it seems py2 is not a valid selector. It should be py2k or py24 / py27. This is from the section on [selectors](https://github.com/conda/conda-docs/blob/master/docs/source/building/meta-yaml.rst#preprocessing-selectors)

**Comment by ajgpitch** [comment]
@sahmed95 hey, thanks for keeping an eye on this and making the update
I'll have a look now to see what's failing.
If it becomes much more hassle, then I am going to ditch the py27 and py34 builds on Windows. It's quite possible they will won't work anyway (like in #16)

**Comment by ajgpitch** [comment]
Looks like in Windows environment vars always have string values, so my conditional compiler choice did not work. I can fix that.
Looks like it did not attempt the py35+ builds on Windows, I guess we need a rerender. I will give this a go.

**Comment by ajgpitch** [comment]
@nonhermitian The Travis (OSX) builds are all failing on compilation, some issue with a `__sort` function. Do you have any ideas on this?

**Comment by ajgpitch** [comment]
@sahmed95 Just out of interest, how did you manage to push on this branch?

**Comment by nonhermitian** [comment]
Since we now use c++,  I switched to the sort function from cython libcpp.  This is now also used in the reverse cuthill mckee function.  I do all dev on OSX, and know that everything builds and all tests pass so not sure off the top of my head why there is an issue.

**Comment by ajgpitch** [comment]
I don't seem to able to rerender. I get this:
```
Fetching package metadata ...........
Traceback (most recent call last):
  File "/home/agp1/anaconda3/bin/conda-smithy", line 6, in <module>
    sys.exit(conda_smithy.cli.main())
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_smithy/cli.py", line 271, in main
    args.subcommand_func(args)
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_smithy/cli.py", line 182, in __call__
    configure_feedstock.main(args.feedstock_directory)
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_smithy/configure_feedstock.py", line 705, in main
    render_circle(env, config, forge_dir)
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_smithy/configure_feedstock.py", line 77, in render_circle
    forge_config.get('channels', {}).get('sources', tuple())
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_smithy/configure_feedstock.py", line 626, in compute_build_matrix
    mtx = special_case_version_matrix(meta, index)
  File "/home/agp1/anaconda3/lib/python3.6/site-packages/conda_build_all/version_matrix.py", line 196, in special_case_version_matrix
    np_vn = minor_vn(index[numpy_pkg.fn]['version'])
KeyError: 'numpy-1.5.1-py26_pro0.tar.bz2'
```
I am using conda-smithy 2.1.1
I pushed some changes anyway, just to see if the Windows py27 and py34 now work.

**Comment by sahmed95** [comment]
@ajgpitch, I was added as a member to the conda-forge organisation and I might have got commit access because of that. I am trying to learn more about conda-forge / smithy etc these days. 

Happy to help :-)

**Comment by ajgpitch** [comment]
@sahmed95 So git tech speaking you just checked out and my branch then pushed a commit onto it? I just did not know this was possible.
Well, clearly we need some expertise on conda-forge. Always seems to be a challenge!

**Comment by sahmed95** [comment]
@ajgpitch It was easier than that. I edited your commit online and pushed it on the same PR. I think you can do that for QuTiP PRs being an admin and having merge access.

**Comment by ajgpitch** [comment]
@sahmed95 Good to know. Thanks

File: recipe/meta.yaml [hunk 1]
```diff
@@ -7,29 +7,28 @@ package:
...
 requirements:
   build:
     - python
     - setuptools
-    - cython >=0.21
     - numpy x.x
+    - cython >=0.21
     - scipy >=0.15
-    - mingwpy                         # [win]
+    - mingwpy                         # [win and (py2k or py34)]
```
Comment (isuruf) [inline recipe/meta.yaml]: Add `toolchain` here

**Comment by nonhermitian** [comment]
It looks like my latest attempt at a fix did not do it.  I am puzzled as every single compiler that I have used seems to compile the code just fine.  GCC 4.7.2 & 4.9.2 & 6.3.0, VS 2015, clang 3.9, and Apple clang 8.0.0 all work.  Not even a warning about the the issues that travis is complaining about here.

**Comment by ajgpitch** [comment]
@nonhermitian I agree, it would seem extremely unlikely that there is anything wrong in the code itself, as it compiles on with all compiler and platform combos you have tried. So it seems we must look to something in the conda-forge build configuration. Perhaps @isuruf's `toolchain` suggestion will fix it, the name gives me hope.

**Comment by isuruf** [comment]
Adding `toolchain` does not fix this particular problem, but it should be added. This seems to be a problem with libc++ which might have been fixed in later versions and that's why clang 3.9 and apple clang 8.0.0 passed and not apple clang 6.1.0.

**Comment by ajgpitch** [comment]
@isuruf I am not able to rerender still though. For now I will hack the appveyor.yml to get the Windows builds we want to see.
I spent quite a long time looking into the conda smithy code, but I could not really see why this key should be missing from the index.

**Comment by isuruf** [comment]
Rerendering works fine for me. Did you try removing and reinstalling conda-smithy?

**Comment by ajgpitch** [comment]
@isuruf I tried deleting and reinstalling anaconda (for py36). I was using one for py27 before, so thought best to update all round. Then I installed conda-smithy
This is the [hopefully] relevant bit of my conda list
```
anaconda                  4.3.1               np111py36_0  
anaconda-client           1.6.0                    py36_0  
anaconda-navigator        1.5.0                    py36_0  
anaconda-project          0.4.1                    py36_0  
conda                     4.3.14                   py36_0  
conda-build               2.1.6                    py36_0  
conda-build-all           1.0.1                    py36_1  
conda-env                 2.6.0                         0  
conda-smithy              2.1.1                    py36_0    conda-forge
conda-verify              2.0.0                    py36_0  
```
Maybe I should force the conda-build-all to take the conda-forge version?

**Comment by isuruf** [comment]
Yes, you can try conda-forge's `conda-build-all`

**Comment by nonhermitian** [comment]
@ajgpitch It is not a tool chain issue, but rather a compiler issue.  It is complaining about some referencing issue that every other compiler seems to not have issue with.  The problem is that in order to hunt it down, I would have to have OSX 10.9 and xcode 6.4, which are both around three years old.

**Comment by ajgpitch** [comment]
I removed conda-smithy and reinstalled with the `-c conda-forge` 
I now have:
```
conda                     4.2.13                   py36_0    conda-forge
conda-build               2.1.6                    py36_0  
conda-build-all           1.0.1                    py36_1  
conda-env                 2.6.0                         0    conda-forge
conda-smithy              2.1.1                    py36_0    conda-forge
conda-verify              2.0.0                    py36_0 
```
and I was able to rerender. Thanks @isuruf

**Comment by isuruf** [comment]
Here's a patch that fixes the particular case, https://github.com/isuruf/qutip-feedstock/blob/04eb924a5c8d62857d8675e690f6c65cddf62a08/recipe/sort.patch

There are more calls to `std::sort` that needs to be fixed.

**Comment by nonhermitian** [comment]
@isuruf Thanks for the patch!  It has been applied.

**Comment by isuruf** [comment]
Yes, python 3.4 was dropped by conda-forge when python 3.6 came out.

**Comment by isuruf** [comment]
@ajgpitch looks like you didn't upload the new distribution.

**Comment by ajgpitch** [comment]
Yes, I missed a merge. Done now. Thanks for spotting @isuruf 
I pushed a rerender, probably a good idea, as I had changed the recipe, but mainly to restart the tests.

**Comment by ajgpitch** [comment]
The good news is that the OSX builds seems to work fine.
The bad news is that all the Windows py27 builds failed today (they were fine yesterday).
The two I checked had this error. This seems like a very strange error.
```
(C:\bld\qutip_1489585700571\_t_env) C:\bld\qutip_1489585700571\test_tmp>if errorlevel 1 exit 1 
(C:\bld\qutip_1489585700571\_t_env) C:\bld\qutip_1489585700571\test_tmp>"C:\bld\qutip_1489585700571\_t_env\python.exe" -s "C:\bld\qutip_1489585700571\test_tmp\run_test.py" 
import: u'qutip'
Traceback (most recent call last):
  File "C:\bld\qutip_1489585700571\test_tmp\run_test.py", line 2, in <module>
    import qutip
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\qutip\__init__.py", line 41, in <module>
    from qutip.utilities import _version2int
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\qutip\utilities.py", line 43, in <module>
    from scipy.misc import factorial
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\scipy\misc\__init__.py", line 53, in <module>
    from scipy.interpolate._pade import pade
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\scipy\interpolate\__init__.py", line 176, in <module>
    from .interpolate import *
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\scipy\interpolate\interpolate.py", line 32, in <module>
    from .interpnd import _ndim_coords_from_arrays
  File "scipy/interpolate/interpnd.pyx", line 1, in init scipy.interpolate.interpnd (scipy\interpolate\interpnd.c:25150)
  File "C:\bld\qutip_1489585700571\_t_env\lib\site-packages\scipy\spatial\__init__.py", line 94, in <module>
    from .ckdtree import *
  File "scipy/spatial/ckdtree.pyx", line 29, in init scipy.spatial.ckdtree (scipy\spatial\ckdtree.cxx:19224)
  File "C:\bld\qutip_1489585700571\_t_env\lib\multiprocessing\__init__.py", line 136, in cpu_count
    raise NotImplementedError('cannot determine number of cpus')
NotImplementedError: cannot determine number of cpus
(C:\bld\qutip_1489585700571\_t_env) C:\bld\qutip_1489585700571\test_tmp>if errorlevel 1 exit 1 
TESTS FAILED: qutip-4.1.0-np112py27_0
Command exited with code 1
```

**Comment by ajgpitch** [comment]
I did another rerender to force a rebuild.
During this morning's rerender we lost the py36 x64 builds from appveyor.yml. The afternoon's rerender  lost the py36 x86 builds from appveyor.yml. @isuruf do we know why these builds have dropped out of the list?

**Comment by nonhermitian** [comment]
For some reason, multiprocessing.cpu_count is raising an exception. It did bot do that before, and I have yet to see it in the wild.  However, a try-except statement in __init__ should do it.

**Comment by ajgpitch** [comment]
Ok, thanks @nonhermitian. I will do some code fix to avoid this problem.

**Comment by nonhermitian** [comment]
The change should set the number of cpus in the QuTiP settings to one if
the try fails.

**Comment by isuruf** [comment]
@ajgpitch, I don't know the reason. For me, when I rerender, it adds the 4 builds.

**Comment by ajgpitch** [comment]
@isuruf I just rerendered and it put them back in, so we'll leave as a mystery not worth solving.

**Comment by nonhermitian** [comment]
It seems that the builds are running into Exceptions with `multiprocessing.cpu_count`.  The call that is in QuTiP can be wrapped in a try-except statement.  However, there is call in `scipy/scipy/spatial/ckdtree.pyx` that is also causing an issue (for us, this is called when loading scipy.misc.factorial).  So the problematic appveyor builds will fail on scipy when anyone needs something from `scipy.misc`.

**Comment by ajgpitch** [comment]
Ok, so I was hoping the win py27 cpu_count() problem might just go away. Seems not.
I will apply qutip/qutip#668 now and trigger a retest now

**Comment by nonhermitian** [comment]
Nope, it is a bug in appveyor.  If a CI suite can't build SciPy then what good is it?

**Comment by ajgpitch** [comment]
@isuruf Should I raise this issue with the AppVeyor people, or is it better coming from the conda-forge core team?

**Comment by ajgpitch** [comment]
the AppVeyor build uses scipy from the default conda channel. I tested using exactly the same conda env here and did not see the exception. Strange
but for now I will be hopeful that the latest source change will handle this issue for us.

**Comment by ajgpitch** [comment]
It would seem that it is too much to ask that qutip is expected to pass any test when `import scipy.misc` fails. For now I will remove the win py27 combination. Perhaps we can try another build at some point in the future when appveyor may have noticed this issue and fixed it

**Comment by ajgpitch** [comment]
I have raised an issue on the appveyor website

**Comment by nonhermitian** [comment]
It seems the recent errors are more due to not being able to download the qutip zip and some ssl issue.

**Comment by ajgpitch** [comment]
The most recent error (before my last commit) was:
```
TEST START: C:\bld\win-64\qutip-4.1.0-np111py27_0.tar.bz2
Deleting work directory, C:\bld\qutip_1489675049058\work\qutip-4.1.0
The following packages will be downloaded:
    package                    |            build
    ---------------------------|-----------------
    qutip-4.1.0                |      np111py27_0         2.2 MB  local
The following NEW packages will be INSTALLED:
    certifi:        2017.1.23-py27_0   conda-forge
    cython:         0.25.2-py27_1      conda-forge
    mingwpy:        0.1.0-py27_2       conda-forge
    mkl:            2017.0.1-0         defaults   
    numpy:          1.11.3-py27_0      defaults   
    python:         2.7.12-2           conda-forge
    qutip:          4.1.0-np111py27_0  local      
    scipy:          0.19.0-np111py27_0 defaults   
    setuptools:     33.1.0-py27_0      conda-forge
    vs2008_runtime: 9.00.30729.5054-0  defaults   
    wincertstore:   0.2-py27_0         conda-forge
WARNING: conda-build appears to be out of date. You have version 2.1.6 but the
latest version is 2.1.7. Run
conda update -n root conda-build
to get the latest version.
C:\bld\qutip_1489675049058\test_tmp>call "C:\Miniconda-x64\Scripts\activate.bat" "C:\bld\qutip_1489675049058\_t_env"  1>nul 2>&1 
(C:\bld\qutip_1489675049058\_t_env) C:\bld\qutip_1489675049058\test_tmp>if errorlevel 1 exit 1 
(C:\bld\qutip_1489675049058\_t_env) C:\bld\qutip_1489675049058\test_tmp>"C:\bld\qutip_1489675049058\_t_env\python.exe" -s "C:\bld\qutip_1489675049058\test_tmp\run_test.py" 
import: u'qutip'
C:\bld\qutip_1489675049058\_t_env\lib\site-packages\qutip\__init__.py:161: UserWarning: matplotlib not found: Graphics will not work.
  warnings.warn("matplotlib not found: Graphics will not work.")
Traceback (most recent call last):
  File "C:\bld\qutip_1489675049058\test_tmp\run_test.py", line 2, in <module>
    import qutip
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\qutip\__init__.py", line 183, in <module>
    from qutip.visualization import *
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\qutip\visualization.py", line 62, in <module>
    from qutip.wigner import wigner
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\qutip\wigner.py", line 51, in <module>
    from scipy.misc import factorial
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\scipy\misc\__init__.py", line 53, in <module>
    from scipy.interpolate._pade import pade
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\scipy\interpolate\__init__.py", line 176, in <module>
    from .interpolate import *
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\scipy\interpolate\interpolate.py", line 32, in <module>
    from .interpnd import _ndim_coords_from_arrays
  File "scipy/interpolate/interpnd.pyx", line 1, in init scipy.interpolate.interpnd (scipy\interpolate\interpnd.c:25150)
  File "C:\bld\qutip_1489675049058\_t_env\lib\site-packages\scipy\spatial\__init__.py", line 94, in <module>
    from .ckdtree import *
  File "scipy/spatial/ckdtree.pyx", line 29, in init scipy.spatial.ckdtree (scipy\spatial\ckdtree.cxx:19224)
  File "C:\bld\qutip_1489675049058\_t_env\lib\multiprocessing\__init__.py", line 136, in cpu_count
    raise NotImplementedError('cannot determine number of cpus')
NotImplementedError: cannot determine number of cpus
(C:\bld\qutip_1489675049058\_t_env) C:\bld\qutip_1489675049058\test_tmp>if errorlevel 1 exit 1 
TESTS FAILED: qutip-4.1.0-np111py27_0
Command exited with code 1
```

**Comment by ajgpitch** [comment]
There was a commit yesterday where I failed to merge the new dist upload, which failed due to openssl sha mismatch.

**Comment by nonhermitian** [comment]
Hell has frozen over.  It works!  Great job @isuruf and @ajgpitch.

**Comment by ajgpitch** [comment]
Yes, it was a bit of a struggle this time.
Hopefully next time there will be no recipe changes and it will all be
quite smooth

**Comment by isuruf** [comment]
Just to make it clear, python 2.7 on windows is skipped. that's why it passes. @ajgpitch, can you do a rerender?

**Comment by ajgpitch** [comment]
@isuruf Yes, I really wanted to get some packages out there. I plan to do another build when I have reason to believe that the issue on appveyor with `cpu_count` on py27 might be fixed. 
I have merged now, can I still rerender and commit to this branch, or should I make another?

**Comment by ajgpitch** [comment]
I am testing packages on linux and windows. All looks good so far.
@nonhermitian @sahmed95  It looks like OSX packages should be available, could you test please.

**Pull request merged**